File: server/app/api/entities/message.rb
module Entities
  class Message < Grape::Entity
    expose :id
    expose :content
    expose :message_type
    expose :created_at

    expose :sender do |message, options|
      Sender.represent(message.user, options)
    end

    expose :is_sender do |message, options|
      message.user_id == options[:current_user_id]
    end
  end

  class Sender < Grape::Entity
    expose :id
    expose :username
    expose :profile do |user, options|
      SenderProfile.represent(user.profile, options)
    end
  end

  class SenderProfile < Grape::Entity
    expose :avatar_settings
  end
end

File: client/src/types/message.ts
export interface CreateMessageResponse {
  id: number;
  message: string;
  chat_room_id: number;
  content: string;
}

export interface CreateMessageRequest {
  receiver_id: string;
  content: string;
  message_type: string;
  chat_room_id: string;
}

export interface Sender {
  id: number;
  username: string;
  profile?: {
    avatarSettings?: {
      avatarContent?: string;
      bgColor?: string;
      textColor?: string;
    };
  };
}

export interface SenderResponse {
  id: number;
  username: string;
  profile?: {
    avatar_settings?: {
      avatar_content?: string;
      bg_color?: string;
      text_color?: string;
    };
  };
}

export interface Message {
  id: number;
  isSender: boolean;
  messageType: string | null;
  sender: Sender;
  content: string;
  createdAt: string;
}

export interface MessageResponse {
  id: number;
  is_sender: boolean;
  message_type: string | null;
  sender: SenderResponse;
  content: string;
  created_at: string;
}
